<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Methods – Monitoring the Rhizosphere Summer ’25 Report</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-5ac60a1f34ed577e3856fa40d0407be7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Monitoring the Rhizosphere Summer ’25 Report</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./people.html"> 
<span class="menu-text">People</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./method.html">Methods &amp; Analysis</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./method.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Methods &amp; Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./results.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Results</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./discussion.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Discussion</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendices</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#arabidopsis-setup" id="toc-arabidopsis-setup" class="nav-link active" data-scroll-target="#arabidopsis-setup">2.1 Arabidopsis Setup</a></li>
  <li><a href="#lci-setup" id="toc-lci-setup" class="nav-link" data-scroll-target="#lci-setup">2.2 LCI Setup</a>
  <ul class="collapse">
  <li><a href="#interferometer-materials" id="toc-interferometer-materials" class="nav-link" data-scroll-target="#interferometer-materials">2.2.1 Interferometer Materials</a></li>
  <li><a href="#lci-hardware" id="toc-lci-hardware" class="nav-link" data-scroll-target="#lci-hardware">2.2.2 LCI Hardware</a></li>
  <li><a href="#lci-software" id="toc-lci-software" class="nav-link" data-scroll-target="#lci-software">2.2.3 LCI Software</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Methods</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>For a list of materials used and their sources see Appendix A.</p>
<section id="arabidopsis-setup" class="level2">
<h2 class="anchored" data-anchor-id="arabidopsis-setup">2.1 Arabidopsis Setup</h2>
<p><em>Arabidopsis thaliana</em> was used as a plant model for integration into the low-coherence interferometry setup. Arabidopsis is a model organism, meaning that the majority of gene families present in the Arabidopsis genome are present in other flowering plants, thus their gene functions are often similar (see <a href="#fig-genes" class="quarto-xref">Figure&nbsp;1</a>). This means that conclusions drawn from research utilizing Arabidopsis as a model organism are often applicable to other flowering plants <span class="citation" data-cites="woodward_biology_2018">(<a href="#ref-woodward_biology_2018" role="doc-biblioref">Woodward and Bartel 2018</a>)</span>.</p>
<div id="fig-genes" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-genes-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="genefamilies.png" class="img-fluid figure-img" style="width:50.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-genes-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: A diagram demonstrating the overlap between gene families of Arabidopsis and other similar flowering plants. Approximately two-thirds of Arabidopsis gene families overlap between all of these plant species <span class="citation" data-cites="paterson_sorghum_2009">(<a href="#ref-paterson_sorghum_2009" role="doc-biblioref">Paterson et al. 2009</a>)</span>.
</figcaption>
</figure>
</div>
<p>Arabidopsis also has a characteristic root structure that integrates well with the interferometry setup, which is a larger, primary root with smaller roots branching off (see <a href="#fig-scan" class="quarto-xref">Figure&nbsp;2</a>).</p>
<div id="fig-scan" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-scan-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="duandinneny.png" class="img-fluid figure-img" style="width:35.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-scan-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: A scanned image of Arabidopsis roots grown in a Standard Murashige and Skoog (MS) medium <span class="citation" data-cites="estevez_salt-stress_2015">(<a href="#ref-estevez_salt-stress_2015" role="doc-biblioref">Duan, Sebastian, and Dinneny 2015</a>)</span>.
</figcaption>
</figure>
</div>
<p>In the low-coherence interferometry setup, a section of the primary root will be used such that the near-infrared beam may pass through the single root. This allows for examination of the light-piping properties of a single Arabidopsis root.</p>
<div id="fig-protocol" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-protocol-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="rootprocess.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-protocol-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: The drawn out protocol for growing the Arabidoposis roots in a wet lab.
</figcaption>
</figure>
</div>
<p>As seen in <a href="#fig-protocol" class="quarto-xref">Figure&nbsp;3</a>, preparing liquid media for plant growth required 0.43 grams of Murashige &amp; Skoog (MS) basal salt media, and 2 grams of sucrose to be mixed into a 25lmL beaker containing 100mL MQ water using a stir bar. Using a pH probe, 0.1M KOH was added until the solution was a pH of 6.0, 4 grams of agar powder, 100mL MQ water, and 0.1M KOH added dropwise for pH adjustment to 6.0, since optimal pH range for plant growth is 6.0-7.5. After pH adjustment, this solution was added to a 150mL Erlenmeyer flask containing 4g of LB agar powder, mixed, and autoclaved for 20 minutes on a liquid cycle.</p>
<p>After sterilization in the autoclave, the 100mLs of liquid media were poured into four clean plates in a sterile environment before being covered with an opaque material and allowed to set and cool overnight. The next morning they were placed upside-down in a refrigerator at 4°C for 24 hours.</p>
<p>Arabidopsis seeds were rinsed in a seed sterilization buffer containing 50μL of 10X Triton (diluted from 100X Triton) and 10mL of 70% ethanol. They were then rinsed twice in MQ water before being plated onto the agar plates. Six seeds were placed on each plate and they were left to germinate at 4°C in the dark for 1-2 days before being incubated in a Vivosun growth chamber at 21-22°C, 65-70% humidity, and a 12 hours light/12 hours dark light cycle. The plants continued to grow on agar plates. The Vivosun growth chamber elements used were a 2x2 growth tent, the AeroStream U075 750ml USB Humidifier, and the AeroWave D4 4” clip fan, as well as a Bluetooth thermometer/hygrometer (see <a href="#fig-tent" class="quarto-xref">Figure&nbsp;4</a>).</p>
<div id="fig-tent" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-tent-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="tentsetup.png" class="img-fluid figure-img" style="width:50.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-tent-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: The plant growth tent setup used to house the seeded agar plates. The setup includes a reflective tent, a fan, an overhanging light, and a humidifier sourced from Vivosun. The agar plate holder was designed on Solidworks and 3D-printed on Harvey Mudd’s makerspace printers.
</figcaption>
</figure>
</div>
</section>
<section id="lci-setup" class="level2">
<h2 class="anchored" data-anchor-id="lci-setup">2.2 LCI Setup</h2>
<section id="interferometer-materials" class="level3">
<h3 class="anchored" data-anchor-id="interferometer-materials">2.2.1 Interferometer Materials</h3>
<div id="fig-lcidrawing" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-lcidrawing-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="lcidrawing.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-lcidrawing-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: A drawing of the Low Coherence Interferometer (LCI) setup with an 850 nm super luminescent diode (SLD) feeding into a polarizer, then split 50:50 by a 2x2 coupler into a sample and reference beam that when reflected recombine into the photodiode (PD). The signal is then fed into a photodiode amplifier (PA), read by the data acquistion device (NiDAQ), and saved on the computer through code.
</figcaption>
</figure>
</div>
<p>As shown in <a href="#fig-lcidrawing" class="quarto-xref">Figure&nbsp;5</a>, the interferometry system used a super luminescent diode (SLD) with a nominal wavelength of 850 nm and maximum power of 20 mW as the broadband low coherence light source. This light was then fed into a 2x2 coupler via an FC/APC fiber that splits the incoming light 50:50 into a sample beam and a reference beam. The sample beam and reference beam were collimated using a fiber collimation package with a Numerical Aperture of 0.15 (NA = 0.15) and a focal length of 18.45 mm. The collimated reference beam enters free space and then is reflected off a total internal reflection (TIR) retroreflector back into the collimator and into the fiber. The collimated sample beam enters free space and is reflected back off the sample back into the fiber to then interfere with the returning reference beam. The combined beam returns to the 2x2 coupler where the beam is split 50:50 once more with half of the beam being sent back to the source and the other half of the beam being sent to a large area SI detector with an active area of 75.4 mm2 to measure the intensity of the interference signal.</p>
<div id="fig-lciphoto" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-lciphoto-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="lcipicture.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-lciphoto-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: A labeled picture of the in-lab LCI setup. The sample in this setup is a TIR retroreflection with an AR coating for 350-700 nm.
</figcaption>
</figure>
</div>
<p>With the following setup pictured in <a href="#fig-lciphoto" class="quarto-xref">Figure&nbsp;6</a>, the detector will only measure a peak in interference when the reference beam is within wavelengths with the light reflected from the sample. The reflected beam from each reflective surface of the sample can then be matched in path length by sweeping the reference mirror/retroreflector. Due to the SLD’s small coherence length of 13 um, the reference mirror needed to be on a programmable translation stage. An extra manual stage was added to the sample and underneath the translation stage holding the retroreflector for ease of alignment. Alignment of the reference retroflector required blocking the path between the output beam from the sample collimator path and the sample, tweaking the position and angle of the retroreflector, and using a power meter in place of the detector in order to visualize and maximize the power output from the reference beam. This process was then repeated for the sample beam by blocking the reference retroreflector instead. Initial sweeps were taken without the Arabidopsis root sample present to find the SLD’s small coherence length before testing with plant sample present.</p>
<p>All materials were sourced from Thorlabs.</p>
</section>
<section id="lci-hardware" class="level3">
<h3 class="anchored" data-anchor-id="lci-hardware">2.2.2 LCI Hardware</h3>
<p>The data is measured as a voltage whereas we are concerned with the incident power onto the photodiode to delineate information about the intensity of the interference signal. The hardware is where the conversion from power to voltage occurs. This section depicts the conversion from the measured voltage back to incident power.</p>
<p>The setup uses a NiDAQ instrument to acquire the data as a measured voltage. The voltage signal (red cable) is connected to analog input AI1 whereas the reference ground (black cable) is connected to analog ground AIGND. The voltage signal and reference ground is collected from the photodiode amplifier.</p>
<p>The PDA200C Photodiode Amplifier (PA) reads the signal from a DET100A2 photodiode (PD) as a current and then outputs a voltage with a set conversion scale.</p>
<section id="pa-voltage-to-pd-current" class="level4">
<h4 class="anchored" data-anchor-id="pa-voltage-to-pd-current">PA Voltage to PD Current</h4>
<p>The analog output of the PA conversion depends on the current range and resolution that is manually set on the PA (see <a href="#fig-conversion" class="quarto-xref">Figure&nbsp;7</a>).</p>
<div id="fig-conversion" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-conversion-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="paconversion.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-conversion-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7: Analog Output Conversion Table from PDA200C datasheet
</figcaption>
</figure>
</div>
<p>For our setup, the PA was set to a resolution of 1 nA for a signal within the 10µA current range, meaning the Conversion Coeff used was 1x10<sup>6</sup> V/A. Using the conversion coefficient and the current range, we can calculate from the output voltage measured what the current seen on the PA is. Connecting to the PA with the PD darkened completely allows the PA to offset the dark current from the photodiode. Thus, the measured voltage can be converted directly to the PD current (I<sub>pd</sub>) without needing to calculate the dark current.</p>
</section>
<section id="pd-current-to-incident-power" class="level4">
<h4 class="anchored" data-anchor-id="pd-current-to-incident-power">PD Current to Incident Power</h4>
<p>The PA can directly convert to power but because the photodiode’s responsivity can change based on light wavelength, it is good to check the power with hand calculations. To calculate power from the PD, we estimate the responsivity from the DET100A2 datasheet for our SLD with wavelength of 850nm to be approximately 0.65.</p>
<p>We can then calculate power using the relationship between responsivity, I<sub>pd</sub>, and power derived from the DET100A2 datasheet (see <a href="#eq-power" class="quarto-xref">Equation&nbsp;1</a>).</p>
<p><span id="eq-power"><span class="math display">\[
P_{incident} = \frac{R(\lambda)}{I_{PD}}    
\tag{1}\]</span></span></p>
</section>
</section>
<section id="lci-software" class="level3">
<h3 class="anchored" data-anchor-id="lci-software">2.2.3 LCI Software</h3>
<p>The Python packages nidaqmx and pylablib were used to connect to the NiDAQ and the Kinetic Piezo motor that translates the stage housing the reference mirror back and forth. Initial data was taken on the NiDAQ with a sampling frequency of 1 MHz with 1000 samples per channel. Given a start and end positional value for the Kinesis Motor, the software program moved the motor from start to end position, pausing at each integer positional value to read all available samples in the NiDAQ channel. These samples would then be averaged and stored as the voltage of the interference signal at that motor position for the trial. For multiple trials, the program swept the reference mirror from start to end position before repeating the cycle from start to end position for the provided number of trials. The program then took the average and the standard deviation of the voltages stored at each position for each trial to produce a list of trial averaged voltages at each position, The program then plotted the trial averaged voltages with the standard deviation as error bars on a graph with the motor position as the independent axis and measured voltage as the dependent axis. With only one trial, the plot would be the averaged voltages of all the read samples for each position for the one trial with no error bars due to a standard deviation of zero.</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>LCISweepCONTINUOUS.py</strong></pre>
</div>
<div class="sourceCode" id="cb1" data-filename="LCISweepCONTINUOUS.py"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1"></a><span class="im">from</span> pylablib.devices <span class="im">import</span> Thorlabs</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="im">import</span> nidaqmx</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="im">from</span> nidaqmx.constants <span class="im">import</span> TerminalConfiguration, AcquisitionType, READ_ALL_AVAILABLE</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="im">import</span> csv</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="im">import</span> os</span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="im">import</span> time</span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="im">from</span> datetime <span class="im">import</span> datetime</span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="im">from</span> scipy.fft <span class="im">import</span> fft, fftfreq</span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="im">from</span> scipy.optimize <span class="im">import</span> curve_fit</span>
<span id="cb1-12"><a href="#cb1-12"></a></span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="kw">def</span> create_folder(folder_name):</span>
<span id="cb1-14"><a href="#cb1-14"></a>    <span class="cf">try</span>:</span>
<span id="cb1-15"><a href="#cb1-15"></a>        os.mkdir(folder_name)</span>
<span id="cb1-16"><a href="#cb1-16"></a>        <span class="bu">print</span>(<span class="ss">f"Folder '</span><span class="sc">{</span>folder_name<span class="sc">}</span><span class="ss">' created successfully."</span>)</span>
<span id="cb1-17"><a href="#cb1-17"></a>    <span class="cf">except</span> <span class="pp">FileExistsError</span>:</span>
<span id="cb1-18"><a href="#cb1-18"></a>        <span class="bu">print</span>(<span class="ss">f"Folder '</span><span class="sc">{</span>folder_name<span class="sc">}</span><span class="ss">' already exists."</span>)</span>
<span id="cb1-19"><a href="#cb1-19"></a>    <span class="cf">except</span> <span class="pp">Exception</span> <span class="im">as</span> e:</span>
<span id="cb1-20"><a href="#cb1-20"></a>        <span class="bu">print</span>(<span class="ss">f"An error occurred: </span><span class="sc">{</span>e<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-21"><a href="#cb1-21"></a></span>
<span id="cb1-22"><a href="#cb1-22"></a><span class="kw">def</span> move_to_folder(folder_name):</span>
<span id="cb1-23"><a href="#cb1-23"></a>    <span class="cf">try</span>:</span>
<span id="cb1-24"><a href="#cb1-24"></a>        os.chdir(folder_name)</span>
<span id="cb1-25"><a href="#cb1-25"></a>        <span class="bu">print</span>(<span class="ss">f"Current directory changed to: </span><span class="sc">{</span>os<span class="sc">.</span>getcwd()<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-26"><a href="#cb1-26"></a>    <span class="cf">except</span> <span class="pp">FileNotFoundError</span>:</span>
<span id="cb1-27"><a href="#cb1-27"></a>        <span class="bu">print</span>(<span class="ss">f"Folder '</span><span class="sc">{</span>folder_name<span class="sc">}</span><span class="ss">' not found."</span>)</span>
<span id="cb1-28"><a href="#cb1-28"></a>    <span class="cf">except</span> <span class="pp">NotADirectoryError</span>:</span>
<span id="cb1-29"><a href="#cb1-29"></a>        <span class="bu">print</span>(<span class="ss">f"'</span><span class="sc">{</span>folder_name<span class="sc">}</span><span class="ss">' is not a directory."</span>)</span>
<span id="cb1-30"><a href="#cb1-30"></a>    <span class="cf">except</span> <span class="pp">Exception</span> <span class="im">as</span> e:</span>
<span id="cb1-31"><a href="#cb1-31"></a>        <span class="bu">print</span>(<span class="ss">f"An error occurred: </span><span class="sc">{</span>e<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-32"><a href="#cb1-32"></a></span>
<span id="cb1-33"><a href="#cb1-33"></a><span class="kw">def</span> create_csv_file(filename, fields, datarows):</span>
<span id="cb1-34"><a href="#cb1-34"></a>    <span class="cf">with</span> <span class="bu">open</span>(filename, <span class="st">'w'</span>) <span class="im">as</span> csvfile:</span>
<span id="cb1-35"><a href="#cb1-35"></a>        <span class="co"># create csv writer object</span></span>
<span id="cb1-36"><a href="#cb1-36"></a>        csvwriter <span class="op">=</span> csv.writer(csvfile, lineterminator<span class="op">=</span><span class="st">'</span><span class="ch">\n</span><span class="st">'</span>)</span>
<span id="cb1-37"><a href="#cb1-37"></a></span>
<span id="cb1-38"><a href="#cb1-38"></a>        <span class="co"># writing the fields</span></span>
<span id="cb1-39"><a href="#cb1-39"></a>        csvwriter.writerow(fields)</span>
<span id="cb1-40"><a href="#cb1-40"></a></span>
<span id="cb1-41"><a href="#cb1-41"></a>        <span class="co"># writing the data rows</span></span>
<span id="cb1-42"><a href="#cb1-42"></a>        csvwriter.writerows(datarows)</span>
<span id="cb1-43"><a href="#cb1-43"></a></span>
<span id="cb1-44"><a href="#cb1-44"></a><span class="kw">def</span> take_data():</span>
<span id="cb1-45"><a href="#cb1-45"></a>    <span class="cf">with</span> nidaqmx.Task() <span class="im">as</span> task:</span>
<span id="cb1-46"><a href="#cb1-46"></a>        task.ai_channels.add_ai_voltage_chan(<span class="st">"Dev1/ai1"</span>, </span>
<span id="cb1-47"><a href="#cb1-47"></a>                                            max_val<span class="op">=</span><span class="dv">10</span>, </span>
<span id="cb1-48"><a href="#cb1-48"></a>                                            terminal_config<span class="op">=</span> TerminalConfiguration.RSE)</span>
<span id="cb1-49"><a href="#cb1-49"></a></span>
<span id="cb1-50"><a href="#cb1-50"></a>        task.timing.cfg_samp_clk_timing(<span class="dv">1000000</span>, </span>
<span id="cb1-51"><a href="#cb1-51"></a>                                        sample_mode<span class="op">=</span>AcquisitionType.FINITE, </span>
<span id="cb1-52"><a href="#cb1-52"></a>                                        samps_per_chan<span class="op">=</span><span class="dv">1000</span>)</span>
<span id="cb1-53"><a href="#cb1-53"></a>        rawdata <span class="op">=</span> task.read(READ_ALL_AVAILABLE)</span>
<span id="cb1-54"><a href="#cb1-54"></a>    <span class="cf">return</span> rawdata</span>
<span id="cb1-55"><a href="#cb1-55"></a></span>
<span id="cb1-56"><a href="#cb1-56"></a><span class="kw">def</span> find_amp(data):</span>
<span id="cb1-57"><a href="#cb1-57"></a>    sorted_data <span class="op">=</span> <span class="bu">sorted</span>(data)</span>
<span id="cb1-58"><a href="#cb1-58"></a></span>
<span id="cb1-59"><a href="#cb1-59"></a>    <span class="co"># calculate minimum average of bottom 100 values</span></span>
<span id="cb1-60"><a href="#cb1-60"></a>    averagemin <span class="op">=</span> np.mean(sorted_data[<span class="dv">0</span>:<span class="dv">99</span>])</span>
<span id="cb1-61"><a href="#cb1-61"></a></span>
<span id="cb1-62"><a href="#cb1-62"></a>    <span class="co"># calculate maximum average of top 100 values</span></span>
<span id="cb1-63"><a href="#cb1-63"></a>    averagemax <span class="op">=</span> np.mean(sorted_data[<span class="bu">len</span>(sorted_data)<span class="op">-</span><span class="dv">100</span>:<span class="bu">len</span>(sorted_data)])</span>
<span id="cb1-64"><a href="#cb1-64"></a></span>
<span id="cb1-65"><a href="#cb1-65"></a>    <span class="co"># calculate amplitude</span></span>
<span id="cb1-66"><a href="#cb1-66"></a>    amplitude <span class="op">=</span> averagemax <span class="op">-</span> averagemin</span>
<span id="cb1-67"><a href="#cb1-67"></a>    <span class="cf">return</span> amplitude</span>
<span id="cb1-68"><a href="#cb1-68"></a></span>
<span id="cb1-69"><a href="#cb1-69"></a><span class="kw">def</span> find_average(matrix):</span>
<span id="cb1-70"><a href="#cb1-70"></a>    <span class="co"># Calculate the average of each row</span></span>
<span id="cb1-71"><a href="#cb1-71"></a>    averages <span class="op">=</span> np.mean(matrix, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb1-72"><a href="#cb1-72"></a>    avg_amplitudes <span class="op">=</span> <span class="bu">list</span>(<span class="bu">map</span>(<span class="bu">float</span>, averages))</span>
<span id="cb1-73"><a href="#cb1-73"></a>    <span class="cf">return</span> avg_amplitudes</span>
<span id="cb1-74"><a href="#cb1-74"></a></span>
<span id="cb1-75"><a href="#cb1-75"></a><span class="kw">def</span> find_std(matrix):</span>
<span id="cb1-76"><a href="#cb1-76"></a>    <span class="co"># Calculate the standard deviation of each row</span></span>
<span id="cb1-77"><a href="#cb1-77"></a>    std_devs <span class="op">=</span> np.std(matrix, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb1-78"><a href="#cb1-78"></a>    std_amplitudes <span class="op">=</span> <span class="bu">list</span>(<span class="bu">map</span>(<span class="bu">float</span>, std_devs))</span>
<span id="cb1-79"><a href="#cb1-79"></a>    <span class="cf">return</span> std_amplitudes</span>
<span id="cb1-80"><a href="#cb1-80"></a></span>
<span id="cb1-81"><a href="#cb1-81"></a><span class="kw">def</span> get_serial(value):</span>
<span id="cb1-82"><a href="#cb1-82"></a>    devices <span class="op">=</span> Thorlabs.list_kinesis_devices()</span>
<span id="cb1-83"><a href="#cb1-83"></a>    device_tuple <span class="op">=</span> devices[value]</span>
<span id="cb1-84"><a href="#cb1-84"></a>    serial <span class="op">=</span> device_tuple[<span class="dv">0</span>]</span>
<span id="cb1-85"><a href="#cb1-85"></a>    <span class="cf">return</span> serial</span>
<span id="cb1-86"><a href="#cb1-86"></a></span>
<span id="cb1-87"><a href="#cb1-87"></a><span class="kw">def</span> get_devices():</span>
<span id="cb1-88"><a href="#cb1-88"></a>    <span class="bu">print</span>(Thorlabs.list_kinesis_devices())</span>
<span id="cb1-89"><a href="#cb1-89"></a></span>
<span id="cb1-90"><a href="#cb1-90"></a><span class="co">## Define class for Kinesis stage</span></span>
<span id="cb1-91"><a href="#cb1-91"></a></span>
<span id="cb1-92"><a href="#cb1-92"></a><span class="kw">class</span> KinesisStage(Thorlabs.KinesisPiezoMotor):</span>
<span id="cb1-93"><a href="#cb1-93"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, serial):</span>
<span id="cb1-94"><a href="#cb1-94"></a>        <span class="bu">super</span>().<span class="fu">__init__</span>(serial)</span>
<span id="cb1-95"><a href="#cb1-95"></a></span>
<span id="cb1-96"><a href="#cb1-96"></a>    <span class="kw">def</span> set_def_channel(<span class="va">self</span>, value):</span>
<span id="cb1-97"><a href="#cb1-97"></a>        <span class="va">self</span>.set_default_channel(value)</span>
<span id="cb1-98"><a href="#cb1-98"></a>        <span class="bu">print</span>(<span class="st">" Default channel:"</span>, value)</span>
<span id="cb1-99"><a href="#cb1-99"></a></span>
<span id="cb1-100"><a href="#cb1-100"></a>    <span class="kw">def</span> move_forward(<span class="va">self</span>, value):</span>
<span id="cb1-101"><a href="#cb1-101"></a>        <span class="va">self</span>.move_by(value)</span>
<span id="cb1-102"><a href="#cb1-102"></a>        time.sleep(<span class="dv">4</span>)</span>
<span id="cb1-103"><a href="#cb1-103"></a>        <span class="bu">print</span>(<span class="st">"Stage moved forward by "</span> <span class="op">+</span> <span class="bu">str</span>(value) <span class="op">+</span> <span class="st">"."</span>)</span>
<span id="cb1-104"><a href="#cb1-104"></a>        position <span class="op">=</span> <span class="va">self</span>.get_position()</span>
<span id="cb1-105"><a href="#cb1-105"></a>        <span class="bu">print</span>(<span class="st">"Current Position: "</span>, position)</span>
<span id="cb1-106"><a href="#cb1-106"></a></span>
<span id="cb1-107"><a href="#cb1-107"></a>    <span class="kw">def</span> move_back(<span class="va">self</span>, value):</span>
<span id="cb1-108"><a href="#cb1-108"></a>        <span class="va">self</span>.move_by(<span class="op">-</span>value)</span>
<span id="cb1-109"><a href="#cb1-109"></a>        time.sleep(<span class="dv">4</span>)</span>
<span id="cb1-110"><a href="#cb1-110"></a>        <span class="bu">print</span>(<span class="st">"Stage moved backward by "</span> <span class="op">+</span> <span class="bu">str</span>((<span class="op">-</span>value)) <span class="op">+</span> <span class="st">"."</span>)</span>
<span id="cb1-111"><a href="#cb1-111"></a>        position <span class="op">=</span> <span class="va">self</span>.get_position()</span>
<span id="cb1-112"><a href="#cb1-112"></a>        <span class="bu">print</span>(<span class="st">"Current Position: "</span>, position)</span>
<span id="cb1-113"><a href="#cb1-113"></a></span>
<span id="cb1-114"><a href="#cb1-114"></a>    <span class="kw">def</span> zero_position(<span class="va">self</span>):</span>
<span id="cb1-115"><a href="#cb1-115"></a>        <span class="va">self</span>.move_to(<span class="dv">0</span>)</span>
<span id="cb1-116"><a href="#cb1-116"></a>        time.sleep(<span class="dv">4</span>)</span>
<span id="cb1-117"><a href="#cb1-117"></a>        <span class="bu">print</span>(<span class="st">"Stage position set to 0."</span>)</span>
<span id="cb1-118"><a href="#cb1-118"></a>        position <span class="op">=</span> <span class="va">self</span>.get_position()</span>
<span id="cb1-119"><a href="#cb1-119"></a>        <span class="bu">print</span>(<span class="st">"Current Position: "</span>, position)</span>
<span id="cb1-120"><a href="#cb1-120"></a></span>
<span id="cb1-121"><a href="#cb1-121"></a></span>
<span id="cb1-122"><a href="#cb1-122"></a><span class="co"># Get serial number of motor</span></span>
<span id="cb1-123"><a href="#cb1-123"></a>serial <span class="op">=</span> get_serial(<span class="dv">0</span>)</span>
<span id="cb1-124"><a href="#cb1-124"></a></span>
<span id="cb1-125"><a href="#cb1-125"></a><span class="co"># Make stage object to move stage</span></span>
<span id="cb1-126"><a href="#cb1-126"></a>stage <span class="op">=</span> KinesisStage(serial)</span>
<span id="cb1-127"><a href="#cb1-127"></a>stage.set_def_channel(<span class="dv">2</span>)</span>
<span id="cb1-128"><a href="#cb1-128"></a></span>
<span id="cb1-129"><a href="#cb1-129"></a><span class="co"># Set up how many test and trials and date</span></span>
<span id="cb1-130"><a href="#cb1-130"></a>testnum <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb1-131"><a href="#cb1-131"></a>trialnum <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb1-132"><a href="#cb1-132"></a>date <span class="op">=</span> datetime.now().strftime(<span class="st">"%m-</span><span class="sc">%d</span><span class="st">-%y"</span>)</span>
<span id="cb1-133"><a href="#cb1-133"></a></span>
<span id="cb1-134"><a href="#cb1-134"></a>test_folder <span class="op">=</span> <span class="st">"Test "</span> <span class="op">+</span> <span class="bu">str</span>(testnum)</span>
<span id="cb1-135"><a href="#cb1-135"></a></span>
<span id="cb1-136"><a href="#cb1-136"></a>create_folder(test_folder)</span>
<span id="cb1-137"><a href="#cb1-137"></a>move_to_folder(test_folder)</span>
<span id="cb1-138"><a href="#cb1-138"></a></span>
<span id="cb1-139"><a href="#cb1-139"></a><span class="co"># Set up initial parameters for collecting sweep data</span></span>
<span id="cb1-140"><a href="#cb1-140"></a>sample_rate <span class="op">=</span> <span class="dv">1000000</span>  <span class="co"># Sample rate in Hz</span></span>
<span id="cb1-141"><a href="#cb1-141"></a>sample_count <span class="op">=</span> <span class="dv">1000</span>  <span class="co"># Number of samples to collect</span></span>
<span id="cb1-142"><a href="#cb1-142"></a></span>
<span id="cb1-143"><a href="#cb1-143"></a><span class="co"># setup lists to store data for each input voltage</span></span>
<span id="cb1-144"><a href="#cb1-144"></a>trial_var <span class="op">=</span> {}</span>
<span id="cb1-145"><a href="#cb1-145"></a>N_x <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb1-146"><a href="#cb1-146"></a>start_pos <span class="op">=</span> <span class="dv">5000000</span></span>
<span id="cb1-147"><a href="#cb1-147"></a>end_pos <span class="op">=</span> start_pos<span class="op">-</span><span class="dv">100000</span><span class="op">-</span>N_x</span>
<span id="cb1-148"><a href="#cb1-148"></a>x <span class="op">=</span> np.arange(start_pos, end_pos, <span class="op">-</span>N_x) </span>
<span id="cb1-149"><a href="#cb1-149"></a></span>
<span id="cb1-150"><a href="#cb1-150"></a><span class="co"># Set the stage to the initial point</span></span>
<span id="cb1-151"><a href="#cb1-151"></a>stage.move_to(start_pos)</span>
<span id="cb1-152"><a href="#cb1-152"></a></span>
<span id="cb1-153"><a href="#cb1-153"></a></span>
<span id="cb1-154"><a href="#cb1-154"></a><span class="co"># setup lists to store fields for csv files</span></span>
<span id="cb1-155"><a href="#cb1-155"></a>test_fields <span class="op">=</span> [<span class="st">'x position'</span>]</span>
<span id="cb1-156"><a href="#cb1-156"></a><span class="cf">for</span> number <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>, trialnum<span class="op">+</span><span class="dv">1</span>):</span>
<span id="cb1-157"><a href="#cb1-157"></a>    test_fields.append(<span class="ss">f'Trial </span><span class="sc">{</span>number<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb1-158"><a href="#cb1-158"></a></span>
<span id="cb1-159"><a href="#cb1-159"></a>test_data <span class="op">=</span> <span class="bu">list</span>(x)</span>
<span id="cb1-160"><a href="#cb1-160"></a></span>
<span id="cb1-161"><a href="#cb1-161"></a>current_position <span class="op">=</span> stage.get_position()</span>
<span id="cb1-162"><a href="#cb1-162"></a></span>
<span id="cb1-163"><a href="#cb1-163"></a>trial_data <span class="op">=</span> []</span>
<span id="cb1-164"><a href="#cb1-164"></a>    </span>
<span id="cb1-165"><a href="#cb1-165"></a><span class="cf">with</span> nidaqmx.Task() <span class="im">as</span> task:</span>
<span id="cb1-166"><a href="#cb1-166"></a>    task.ai_channels.add_ai_voltage_chan(<span class="st">"Dev1/ai1"</span>, </span>
<span id="cb1-167"><a href="#cb1-167"></a>                                        max_val<span class="op">=</span><span class="dv">10</span>, </span>
<span id="cb1-168"><a href="#cb1-168"></a>                                        terminal_config<span class="op">=</span> TerminalConfiguration.RSE)</span>
<span id="cb1-169"><a href="#cb1-169"></a></span>
<span id="cb1-170"><a href="#cb1-170"></a>    task.timing.cfg_samp_clk_timing(<span class="dv">1000000</span>, </span>
<span id="cb1-171"><a href="#cb1-171"></a>                                    sample_mode<span class="op">=</span>AcquisitionType.CONTINUOUS, </span>
<span id="cb1-172"><a href="#cb1-172"></a>                                    samps_per_chan<span class="op">=</span><span class="dv">50000</span>)</span>
<span id="cb1-173"><a href="#cb1-173"></a>    </span>
<span id="cb1-174"><a href="#cb1-174"></a>    task.start()</span>
<span id="cb1-175"><a href="#cb1-175"></a>    <span class="bu">print</span>(<span class="st">'Task started.    Data acquiring...'</span>)</span>
<span id="cb1-176"><a href="#cb1-176"></a>    <span class="cf">while</span> current_position <span class="op">&gt;</span> end_pos:</span>
<span id="cb1-177"><a href="#cb1-177"></a>        stage.jog(<span class="st">'-'</span>)</span>
<span id="cb1-178"><a href="#cb1-178"></a>        rawdata <span class="op">=</span> task.read(number_of_samples_per_channel<span class="op">=</span><span class="dv">100</span>)</span>
<span id="cb1-179"><a href="#cb1-179"></a>        rawdata_avg <span class="op">=</span> np.average(rawdata)</span>
<span id="cb1-180"><a href="#cb1-180"></a>        trial_data.append(rawdata_avg)</span>
<span id="cb1-181"><a href="#cb1-181"></a>    task.stop()</span>
<span id="cb1-182"><a href="#cb1-182"></a>    <span class="bu">print</span>(<span class="st">'Task stopped.'</span>)</span>
<span id="cb1-183"><a href="#cb1-183"></a></span>
<span id="cb1-184"><a href="#cb1-184"></a></span>
<span id="cb1-185"><a href="#cb1-185"></a><span class="co"># time.sleep(0.5)</span></span>
<span id="cb1-186"><a href="#cb1-186"></a></span>
<span id="cb1-187"><a href="#cb1-187"></a><span class="co"># rawdata = take_data()</span></span>
<span id="cb1-188"><a href="#cb1-188"></a></span>
<span id="cb1-189"><a href="#cb1-189"></a></span>
<span id="cb1-190"><a href="#cb1-190"></a></span>
<span id="cb1-191"><a href="#cb1-191"></a>test_data <span class="op">=</span> np.vstack([test_data, trial_data])</span>
<span id="cb1-192"><a href="#cb1-192"></a>stage.stop()</span>
<span id="cb1-193"><a href="#cb1-193"></a></span>
<span id="cb1-194"><a href="#cb1-194"></a><span class="co"># Create a csv file with the position and all the trial data</span></span>
<span id="cb1-195"><a href="#cb1-195"></a>testfilename <span class="op">=</span> date <span class="op">+</span> <span class="st">'_ALLTRIALSLCISweep_Test'</span> <span class="op">+</span> <span class="bu">str</span>(testnum) <span class="op">+</span> <span class="st">'.csv'</span></span>
<span id="cb1-196"><a href="#cb1-196"></a>test_datarows <span class="op">=</span> np.transpose(test_data)</span>
<span id="cb1-197"><a href="#cb1-197"></a>create_csv_file(testfilename, test_fields, test_datarows)</span>
<span id="cb1-198"><a href="#cb1-198"></a></span>
<span id="cb1-199"><a href="#cb1-199"></a><span class="co">## Find average of all trials</span></span>
<span id="cb1-200"><a href="#cb1-200"></a></span>
<span id="cb1-201"><a href="#cb1-201"></a>test_data_nox <span class="op">=</span> test_datarows[:,<span class="dv">1</span>:]         <span class="co"># removes x position column</span></span>
<span id="cb1-202"><a href="#cb1-202"></a>averages <span class="op">=</span> find_average(test_data_nox)</span>
<span id="cb1-203"><a href="#cb1-203"></a>std_devs <span class="op">=</span> find_std(test_data_nox)</span>
<span id="cb1-204"><a href="#cb1-204"></a>avg_test_data <span class="op">=</span> np.transpose(np.vstack([<span class="bu">list</span>(x),averages,std_devs]))</span>
<span id="cb1-205"><a href="#cb1-205"></a></span>
<span id="cb1-206"><a href="#cb1-206"></a>avgdatafields <span class="op">=</span> [<span class="st">'x position'</span>, <span class="st">'Average Vout [V]'</span>, <span class="st">'Standard Dev [V]'</span>]</span>
<span id="cb1-207"><a href="#cb1-207"></a>avgdataname <span class="op">=</span> date <span class="op">+</span> <span class="st">'_AveragedLCISweep_Test'</span> <span class="op">+</span> <span class="bu">str</span>(testnum) </span>
<span id="cb1-208"><a href="#cb1-208"></a>avgdatafilename <span class="op">=</span> avgdataname <span class="op">+</span> <span class="st">'.csv'</span></span>
<span id="cb1-209"><a href="#cb1-209"></a>create_csv_file(avgdatafilename, avgdatafields, avg_test_data)</span>
<span id="cb1-210"><a href="#cb1-210"></a></span>
<span id="cb1-211"><a href="#cb1-211"></a><span class="co"># save plot of data for each trial and vin</span></span>
<span id="cb1-212"><a href="#cb1-212"></a>plt.figure()</span>
<span id="cb1-213"><a href="#cb1-213"></a>plt.errorbar(<span class="bu">list</span>(x), averages, yerr<span class="op">=</span>std_devs, fmt<span class="op">=</span><span class="st">'o'</span>)</span>
<span id="cb1-214"><a href="#cb1-214"></a>plt.xlabel(<span class="st">"x position"</span>)</span>
<span id="cb1-215"><a href="#cb1-215"></a>plt.ylabel(<span class="st">'Voltage [V]'</span>)</span>
<span id="cb1-216"><a href="#cb1-216"></a>plt.title(<span class="st">"Test "</span> <span class="op">+</span> <span class="bu">str</span>(testnum) <span class="op">+</span> <span class="st">": LCI Sweep Data"</span>)</span>
<span id="cb1-217"><a href="#cb1-217"></a>plt.savefig(avgdataname <span class="op">+</span><span class="st">'.png'</span>)</span>
<span id="cb1-218"><a href="#cb1-218"></a><span class="co"># plt.show(block=True)</span></span>
<span id="cb1-219"><a href="#cb1-219"></a></span>
<span id="cb1-220"><a href="#cb1-220"></a>os.chdir(<span class="st">'..'</span>)        </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>




</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-estevez_salt-stress_2015" class="csl-entry" role="listitem">
Duan, Lina, Jose Sebastian, and Jose R. Dinneny. 2015. <span>“Salt-<span>Stress</span> <span>Regulation</span> of <span>Root</span> <span>System</span> <span>Growth</span> and <span>Architecture</span> in <span>Arabidopsis</span> <span>Seedlings</span>.”</span> In <em>Plant <span>Cell</span> <span>Expansion</span></em>, edited by José M. Estevez, 1242:105–22. New York, NY: Springer New York. <a href="https://doi.org/10.1007/978-1-4939-1902-4_10">https://doi.org/10.1007/978-1-4939-1902-4_10</a>.
</div>
<div id="ref-paterson_sorghum_2009" class="csl-entry" role="listitem">
Paterson, Andrew H., John E. Bowers, Rémy Bruggmann, Inna Dubchak, Jane Grimwood, Heidrun Gundlach, Georg Haberer, et al. 2009. <span>“The <span>Sorghum</span> Bicolor Genome and the Diversification of Grasses.”</span> <em>Nature</em> 457 (7229): 551–56. <a href="https://doi.org/10.1038/nature07723">https://doi.org/10.1038/nature07723</a>.
</div>
<div id="ref-woodward_biology_2018" class="csl-entry" role="listitem">
Woodward, Andrew W, and Bonnie Bartel. 2018. <span>“Biology in <span>Bloom</span>: <span>A</span> <span>Primer</span> on the <em><span>Arabidopsis</span> Thaliana</em> <span>Model</span> <span>System</span>.”</span> <em>Genetics</em> 208 (4): 1337–49. <a href="https://doi.org/10.1534/genetics.118.300755">https://doi.org/10.1534/genetics.118.300755</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // default icon
          link.classList.add("external");
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>